packages/workshop-app/app/routes/_app+/_exercises+/$exerciseNumber_.$stepNumber.$type.tsx

```diff
@@ -348,6 +375,86 @@ export default function ExercisePartRoute() {
 	const inBrowserBrowserRef = useRef<InBrowserBrowserRef>(null)
 	const previewAppUrl = data.playground?.dev.baseUrl
 
+	const DiffLink = useMemo(() => {
+		return function DiffLink({
+			app1,
+			app2,
+			children,
+			preview,
+			to,
+		}: {
+			app1?: string | number | null
+			app2?: string | number | null
+			to?: string
+			preview?: boolean
+			children?: React.ReactNode
+		}) {
+			if (!to && !app1 && !app2) {
+				return (
+					<callout-danger className="notification">
+						<div className="title">DiffLink Error: invalid input</div>
+					</callout-danger>
+				)
+			}
+
+			function getAppName(input: typeof app1) {
+				if (typeof input === 'number') {
+					const stepIndex = data.exerciseIndex + input
+					return data.allApps[stepIndex]?.name
+				}
+				if (!input) return null
+				for (const { name, stepName } of data.allApps) {
+					if (input === name || input === stepName) {
+						return name
+					}
+				}
+				return null
+			}
+
+			if (to) {
+				const params = new URLSearchParams(to)
+				app1 = params.get('app1')
+				app2 = params.get('app2')
+			}
+			const app1Name = getAppName(app1)
+			const app2Name = getAppName(app2)
+			if (!app1Name || !app2Name) {
+				return (
+					<callout-danger className="notification">
+						<div className="title">DiffLink Error: invalid input</div>
+						{!app1Name && <div>app1: "{app1}" is not a valid app name</div>}
+						{!app2Name && <div>app2: "{app2}" is not a valid app name</div>}
+					</callout-danger>
+				)
+			}
+
+			if (!to) {
+				to = `app1=${app1Name}&app2=${app2Name}`
+			}
+			const pathToDiff = preview
+				? `?${decodeURIComponent(
+						withParam(
+							new URLSearchParams(),
+							'preview',
+							`diff&${to}`,
+						).toString(),
+				  )}`
+				: `/diff?${to}`
+
+			if (!children) {
+				const msg = (s: string) => s.split('__sep__')[2] ?? ''
+				children = (
+					<div>
+						Go to Diff {preview ? 'Preview' : ''} from:{' '}
+						<code>{msg(app1Name)}</code> to: <code>{msg(app2Name)}</code>
+					</div>
+				)
```

Comment:
I want to make sure I can use `DiffLink` inline, and I think this `div` won't work like this.

So I want to be able to say:

````
If you want, you can <Diff app1={-1} app2={0}>the diff with the previous step</Diff> to see the changes that were made since you were last here.
````

And that should all appear on a single line.

Reply:
![image](https://github.com/epicweb-dev/kcdshop/assets/3650909/3e8c1cea-f130-4fbb-b259-3cde9017de13)

i will change it to span